core/src/main/scala/kafka/server/FetchSession.scala

```diff
@@ -0,0 +1,716 @@
+    * @param now        The current time in milliseconds.
+    * @return           True if an entry was evicted; false otherwise.
+    */
+  def tryEvict(privileged: Boolean, key: EvictableKey, now: Long) = synchronized {
+    // Try to evict an entry which is stale.
+    val lastUsedEntry = lastUsed.firstEntry()
+    if (lastUsedEntry == null) {
+      logger.debug("There are no cache entries to evict.")
+      false
+    } else if (now - lastUsedEntry.getKey().lastUsedMs > evictionMs) {
+      val session = lastUsedEntry.getValue()
+      logger.debug("Evicting stale FetchSession {}.", session.id)
+      remove(session)
+      evictionsMeter.mark()
+      true
+    } else {
+      // If there are no stale entries, check the first evictable entry.
+      // If it is less valuable than our proposed entry, evict it.
+      val map = if (privileged) evictableByPrivileged else evictableByAll
+      val evictableEntry = map.firstEntry()
+      if (evictableEntry == null) {
+        logger.debug("No evictable entries found.")
+        false
+      } else if (key.compareTo(evictableEntry.getKey()) < 0) {
+        logger.debug("Can't evict {} with {}.", evictableEntry.getKey(), key)
+        false
+      } else {
+        logger.debug("Evicting {} with {}.", evictableEntry.getKey(), key)
+        remove(evictableEntry.getValue())
+        evictionsMeter.mark()
+        true
+      }
+    }
+  }
+
+  def remove(sessionId: Int): Option[FetchSession] = synchronized {
+    get(sessionId) match {
+      case None => None
+      case Some(session) => remove(session)
+    }
+  }
+
+  /**
+    * Remove an entry from the session cache.
+    *
+    * @param session  The session.
+    *
+    * @return         The removed session, or None if there was no such session.
+    */
+  def remove(session: FetchSession): Option[FetchSession] = synchronized {
+    val evictableKey = session.synchronized {
+      lastUsed.remove(session.lastUsedKey())
+      session.evictableKey()
+    }
+    evictableByAll.remove(evictableKey)
+    evictableByPrivileged.remove(evictableKey)
+    val removeResult = sessions.remove(session.id)
+    if (removeResult.isDefined) {
+      numPartitions = numPartitions - session.cachedSize
+    }
+    removeResult
+  }
+
+  /**
+    * Update a session's position in the lastUsed and evictable trees.
+    *
+    * @param session  The session.
+    * @param now      The current time in milliseconds.
+    */
+  def touch(session: FetchSession, now: Long) = synchronized {
+    session.synchronized {
+      // Update the lastUsed map.
+      lastUsed.remove(session.lastUsedKey())
+      session.lastUsedMs = now
+      lastUsed.put(session.lastUsedKey(), session)
+
+      val oldSize = session.cachedSize
+      if (oldSize != -1) {
+        evictableByPrivileged.remove(session.evictableKey)
+        evictableByAll.remove(session.evictableKey)
+        numPartitions = numPartitions - oldSize
```

Comment:
It's also used in `FetchManager#newContext`.  It would be awkward to not have it since it would mean that there was no good way to remove an arbitrary session (unless you had extra information about the session that wasn't contained in the session itself)